mirrord/agent/src/main.rs

```diff
@@ -66,27 +69,42 @@ mod outgoing;
 mod runtime;
 mod sniffer;
 mod steal;
+mod tls_verifier;
 mod util;
 mod watched_task;
 
+/// Size of [`mpsc`] channels connecting [`TcpStealerApi`] and [`TcpSnifferApi`] with their
+/// background tasks.
 const CHANNEL_SIZE: usize = 1024;
 
-/// Keeps track of connected clients.
-#[derive(Debug)]
+/// Keeps track of next client id.
+/// Stores common data used when serving client connections.
+/// Can be cheaply cloned and passed to per-client background tasks.
+#[derive(Clone)]
 struct State {
-    clients: HashSet<ClientId>,
-    index_allocator: IndexAllocator<ClientId, 100>,
+    /// [`ClientId`] for the next client that connects to this agent.
+    next_client_id: Arc<AtomicU32>,
     /// Handle to the target container if there is one.
     /// This is optional because it is acceptable not to pass the container runtime and id if not
     /// pausing. When those args are not passed, container is [`None`].
     container: Option<ContainerHandle>,
-    env: HashMap<String, String>,
+    env: Arc<HashMap<String, String>>,
     ephemeral: bool,
+    tls_connector: Option<TlsConnector>,
 }
 
 impl State {
     /// Return [`Err`] if container runtime operations failed.
     pub async fn new(args: &Args, watch: drain::Watch) -> Result<State> {
+        let tls_connector = args.use_tls.then(|| {
+            TlsConnector::from(Arc::new(
+                ClientConfig::builder()
+                    .dangerous()
```

Comment:
:skull_and_crossbones: 

This is why I wish rust allowed emojis as identifiers, just imagine, _fn_ :skull_and_crossbones:.

Reply:
:skull_and_crossbones: **danger** :skull_and_crossbones: